src/libpriv/rpmostree-bwrap.c

```diff
@@ -92,50 +145,144 @@ child_setup_fchdir (gpointer user_data)
+                                     "--ro-bind", "/sys/bus", "/sys/bus",
+                                     "--ro-bind", "/sys/class", "/sys/class",
+                                     "--ro-bind", "/sys/dev", "/sys/dev",
+                                     "--ro-bind", "/sys/devices", "/sys/devices",
+                                     NULL);
+
+  for (guint i = 0; i < G_N_ELEMENTS (usr_links); i++)
+    {
+      const char *subdir = usr_links[i];
+      struct stat stbuf;
+      g_autofree char *srcpath = NULL;
+      g_autofree char *destpath = NULL;
+
+      if (fstatat (rootfs_fd, subdir, &stbuf, AT_SYMLINK_NOFOLLOW) < 0)
+        {
+          if (errno != ENOENT)
+            {
+              glnx_set_error_from_errno (error);
+              return FALSE;
+            }
+          continue;
+        }
+      else if (!S_ISLNK (stbuf.st_mode))
+        continue;
+
+      srcpath = g_strconcat ("usr/", subdir, NULL);
+      destpath = g_strconcat ("/", subdir, NULL);
+      rpmostree_bwrap_append_bwrap_argv (ret, "--symlink", srcpath, destpath, NULL);
+    }
+
+  switch (mutable)
+    {
+    case RPMOSTREE_BWRAP_IMMUTABLE:
+      rpmostree_bwrap_append_bwrap_argv (ret, "--ro-bind", "usr", "/usr", NULL);
+      break;
+    case RPMOSTREE_BWRAP_MUTATE_ROFILES:
+      if (!setup_rofiles_usr (ret, error))
+        goto out;
+      break;
+    case RPMOSTREE_BWRAP_MUTATE_FREELY:
+      rpmostree_bwrap_append_bwrap_argv (ret, "--bind", "usr", "/usr", NULL);
+      break;
+    }
+
+  { const char *arg;
+    va_start (args, error);
+    while ((arg = va_arg (args, char *)))
+      g_ptr_array_add (ret->argv, g_strdup (arg));
+    va_end (args);
+  }
+
+  retval = g_steal_pointer (&ret);
+ out:
+  return retval;
+}
 
 static void
 bwrap_child_setup (gpointer data)
 {
-  ChildSetupData *cdata = data;
+  RpmOstreeBwrap *bwrap = data;
 
-  if (fchdir (cdata->rootfs_fd) < 0)
+  if (fchdir (bwrap->rootfs_fd) < 0)
     err (1, "fchdir");
 
-  if (cdata->func)
-    cdata->func (cdata->data);
+  if (bwrap->child_setup_func)
+    bwrap->child_setup_func (bwrap->child_setup_data);
+}
+
+void
+rpmostree_bwrap_set_child_setup (RpmOstreeBwrap *bwrap,
+                                 GSpawnChildSetupFunc func,
+                                 gpointer             data)
+{
+  g_assert (!bwrap->executed);
+  bwrap->child_setup_func = func;
+  bwrap->child_setup_data = data;
 }
 
 gboolean
-rpmostree_run_bwrap_sync_setup (char **argv_array,
-                                int rootfs_fd,
-                                GSpawnChildSetupFunc func,
-                                gpointer             data,
-                                GError **error)
+rpmostree_bwrap_run (RpmOstreeBwrap *bwrap,
+                     GError **error)
 {
   int estatus;
   const char *current_lang = getenv ("LANG");
-  ChildSetupData csetupdata = { rootfs_fd, func, data };
+
+  g_assert (!bwrap->executed);
```

Comment:
I'm increasingly less of a fan of that...it tends to get ignored.  And often there isn't a lot of thought to how the process behaves if a precondition is violated - things are worse if you e.g. return a NULL pointer to some place that isn't expecting it and most of the time SEGV, if you're really unlucky end up executing arbitrary other code.  Better to crash.